chain_of_trust.go

```diff
@@ -0,0 +1,172 @@
+package main
+
+import (
+	"bytes"
+	"crypto/sha256"
+	"encoding/hex"
+	"encoding/json"
+	"io"
+	"os"
+	"path/filepath"
+
+	"golang.org/x/crypto/openpgp"
+
+	"github.com/taskcluster/taskcluster-client-go/queue"
+)
+
+type ChainOfTrust struct {
+}
+
+type ArtifactHash struct {
+	Name string `json:"name"`
+	Hash string `json:"hash"`
+}
+
+type CotExtra struct {
+	PublicIPAddress  string `json:"publicIpAddress"`
+	PrivateIPAddress string `json:"privateIpAddress"`
+	InstanceID       string `json:"instanceId"`
+	InstanceType     string `json:"instanceType"`
+	Region           string `json:"region"`
+}
+
+type ChainOfTrustCertificate struct {
+	Artifacts   []ArtifactHash               `json:"artifacts"`
+	Task        queue.TaskDefinitionResponse `json:"task"`
+	TaskID      string                       `json:"taskId"`
+	RunID       uint                         `json:"runId"`
+	WorkerGroup string                       `json:"workerGroup"`
+	WorkerID    string                       `json:"workerId"`
+	Extra       CotExtra                     `json:"extra"`
+}
+
+func (cot *ChainOfTrust) Initialise() error {
+	return nil
+}
+
+func (cot *ChainOfTrust) IsEnabled(fl EnabledFeatures) bool {
+	return fl.GenerateCertificate
+}
+
+func (cot *ChainOfTrust) Killed(task *TaskRun) error {
+	logFile := filepath.Join(TaskUser.HomeDir, "public", "logs", "live_backing.log")
+	certifiedLogFile := filepath.Join(TaskUser.HomeDir, "public", "logs", "certified.log")
+	signedCert := filepath.Join(TaskUser.HomeDir, "public", "logs", "certified.log.gpg")
```

Comment:
I agree your solution has the advantage of isolating the path to one location, but I also see it as having the disadvantage of adding indirection to the code (it is harder to read at a glance where the vars point to). Since the lines are located together, it is easy to edit them at once. I used to prefer to remove any possible source of duplication, but it always needs to be balanced with whether it makes the code less readable. Since text searching in source code is so easy, in this code I prefer to be explicit. Also, there is no guarantee that we'd always want them to all be located under `public/logs` so if one of them moved somewhere else, you'd be required to refactor the code again. This way, you don't have to, so I feel it protects somewhat against future changes, rather than being a hindrance. :-)